distribution/pull_v2.go

```diff
@@ -98,91 +92,101 @@ func (p *v2Puller) pullV2Repository(ref reference.Named) (err error) {
+	digest         digest.Digest
+	repo           distribution.Repository
+	blobSumService *metadata.BlobSumService
 }
 
-type errVerification struct{}
+func (ld *v2LayerDescriptor) Key() string {
+	return "v2:" + ld.digest.String()
+}
 
-func (errVerification) Error() string { return "verification failed" }
+func (ld *v2LayerDescriptor) ID() string {
+	return stringid.TruncateID(ld.digest.String())
+}
 
-func (p *v2Puller) download(di *downloadInfo) {
-	logrus.Debugf("pulling blob %q", di.digest)
+func (ld *v2LayerDescriptor) DiffID() (layer.DiffID, error) {
+	return ld.blobSumService.GetDiffID(ld.digest)
+}
+
+func (ld *v2LayerDescriptor) Download(ctx context.Context, progressChan chan<- xfer.Progress) (io.ReadCloser, int64, error) {
+	logrus.Debugf("pulling blob %q", ld.digest)
 
-	blobs := p.repo.Blobs(context.Background())
+	blobs := ld.repo.Blobs(ctx)
 
-	desc, err := blobs.Stat(context.Background(), di.digest)
+	desc, err := blobs.Stat(ctx, ld.digest)
 	if err != nil {
 		logrus.Debugf("Error statting layer: %v", err)
-		di.err <- err
-		return
+		if err == distribution.ErrBlobUnknown {
+			return nil, 0, xfer.DoNotRetry{err}
+		}
+		return nil, 0, err
 	}
-	di.size = desc.Size
 
-	layerDownload, err := blobs.Open(context.Background(), di.digest)
+	layerDownload, err := blobs.Open(ctx, ld.digest)
 	if err != nil {
 		logrus.Debugf("Error fetching layer: %v", err)
-		di.err <- err
-		return
+		return nil, 0, err
 	}
-	defer layerDownload.Close()
 
-	verifier, err := digest.NewDigestVerifier(di.digest)
+	reader := xfer.NewProgressReader(ioutils.NewCancelReadCloser(ctx, layerDownload), progressChan, desc.Size, ld.ID(), "Downloading")
+	defer reader.Close()
+
+	verifier, err := digest.NewDigestVerifier(ld.digest)
 	if err != nil {
-		di.err <- err
-		return
+		return nil, 0, err
 	}
 
-	digestStr := di.digest.String()
+	tmpFile, err := ioutil.TempFile("", "GetImageBlob")
+	if err != nil {
+		return nil, 0, err
+	}
 
-	reader := progressreader.New(progressreader.Config{
-		In:        ioutil.NopCloser(io.TeeReader(layerDownload, verifier)),
-		Out:       di.broadcaster,
-		Formatter: p.sf,
-		Size:      di.size,
-		NewLines:  false,
-		ID:        stringid.TruncateID(digestStr),
-		Action:    "Downloading",
-	})
-	io.Copy(di.tmpFile, reader)
+	_, err = io.Copy(tmpFile, io.TeeReader(reader, verifier))
+	if err != nil {
+		return nil, 0, err
+	}
 
-	di.broadcaster.Write(p.sf.FormatProgress(stringid.TruncateID(digestStr), "Verifying Checksum", nil))
+	progressChan <- xfer.Progress{ID: ld.ID(), Action: "Verifying Checksum"}
 
 	if !verifier.Verified() {
-		err = fmt.Errorf("filesystem layer verification failed for digest %s", di.digest)
+		err = fmt.Errorf("filesystem layer verification failed for digest %s", ld.digest)
 		logrus.Error(err)
-		di.err <- err
-		return
+		tmpFile.Close()
+		if err := os.RemoveAll(tmpFile.Name()); err != nil {
+			logrus.Errorf("Failed to remove temp file: %s", tmpFile.Name())
+		}
+
+		return nil, 0, err
```

Comment:
Yes. The thinking is that if we receive something that doesn't match the digest, it could have been corrupted in transit, for example by a misbehaving proxy, or a broken server in a load balancing rotation. Retrying has a chance of getting the expected content.

There is also an argument to be made that if we don't get the right data the first time, we aren't likely to get it after a retry either.

I'm not sure what makes most sense in this case.